cores/nRF5/Tone.cpp

```diff
@@ -34,161 +34,397 @@ Version Modified By Date     Comments
+* by calling noTone().
+* 
+* What is available to configure looping with minimal memory usage:
+* 1. SEQ[n].REFRESH === how many PWM period before loading next sample from sequence
+*                       Thus, setting to 99 will cause 100 pulses per item
+*                       Treat as a 23-bit value.
+* 2. LOOP           === how many times to loop back to SEQ[0]
+*                       SEQ[0] will play the same count if started PWM at SEQ[0]
+*                       SEQ[0] will play one fewer times if started PWM at SEQ[1]
+*                       Treat as a 15-bit value.
+*
+* Therefore, between REFRESH and LOOP, can support up to 40-bit repeat WITHOUT INTERRUPTS.
+*
+* The value of duration is given in milliseconds.
+* Frequency is limited to range [20 ... 25000].
+* 
+* Therefore, maximum pulse count is limited as follows:
+*     (32-bit duration) * (20 ... 25000) / 1000UL
+*     (0xFFFF_FFFF) * 25
+*     0x18_FFFF_FFE7 ... which is 37 bits
+* 
+* Therefore, all possible values for tone() can be supported
+* via a single one-time configuration of the PWM peripheral.
+*
+* PWM peripheral can be de-allocated by sketch call to noTone().
+* 
+* Design:
+* 0. At each call to tone(), unconditionally stop any existing playback.
+* 1. For infinite duration, configure large REFRESH and LOOP (to minimize reading of RAM / AHB traffic),
+*    and setup shortcut to repeat infinitely.
+* 2. For specified duration, configure both SEQ0 and SEQ1:
+*    1a. SEQ[1].REFRESH <-- total count % _iterations_per_loop
+*    1b. SEQ[0].REFRESH <-- _iterations_per_loop - SEQ[1].CNT
+*    1c. LOOP           <-- duration_count / _iterations_per_loop
+* 
+* Result: Zero CPU usage, minimal AHB traffic
+*/
 void tone(uint8_t pin, unsigned int frequency, unsigned long duration)
 {
-	unsigned int time_per=0;
-	
-	if((frequency < 20) | (frequency > 25000)) return;
-	
-
-	float per=float(1)/frequency;
-	time_per=per/0.000008;
-	unsigned int duty=time_per/2;
-	if(duration > 0){
-		no_stop = false;
-		float mil=float(duration)/1000;
-		if(per>mil)
-			count_duration=1;
-		else
-			count_duration= mil/per;
-	}
-	else
-		no_stop = true;
-
-	// Configure PWM
-	static uint16_t seq_values[]={0};
-	//In each value, the most significant bit (15) determines the polarity of the output
-	//0x8000 is MSB = 1
-	seq_values[0]= duty | 0x8000;
-	nrf_pwm_sequence_t const seq={
-								seq_values,
-								NRF_PWM_VALUES_LENGTH(seq_values),
-								0,
-								0
+    static_assert(sizeof(unsigned long int) <= sizeof(uint32_t));
+    static_assert(sizeof(unsigned      int) <= sizeof(uint32_t));
+
+    // Used only to protect calls against simultaneous multiple calls to tone().
+    // Using a function-local static to avoid accidental reference from ISR or elsewhere,
+    // and to simplify ensuring the semaphore gets initialized.
+    static StaticSemaphore_t _tone_semaphore_allocation;
+    static auto init_semaphore = [] () throw() { //< use a lambda to both initialize AND give the mutex
```

Comment:
note for semaphore usage, if you need mutex acess, please use mutex https://www.freertos.org/xSemaphoreCreateMutexStatic.html.

Reply:
This was written as a semaphore because, initially, there was the possibility that synchronization with the ISR would be needed.  The ISR is now gone, so... Yes, I can change this to a mutex.

Reply:
Commit a566c693d9fbea5f97137b8a92e36b76cb05226e